src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java

```diff
@@ -171,7 +147,107 @@ public void handleException(TransportException exp) {
+                publishResponseHandler.onFailure(node, e);
+                return;
+            }
+        }
+        publishClusterStateToNode(clusterState, bytes, node, timedOutWaitingForNodes, publishTimeout, publishResponseHandler);
+    }
+
+    private void sendClusterStateDiff(ClusterState clusterState, Diff diff, Map<Version, BytesReference> serializedDiffs, DiscoveryNode node,
+                                      AtomicBoolean timedOutWaitingForNodes, TimeValue publishTimeout,
+                                      BlockingClusterStatePublishResponseHandler publishResponseHandler) {
+        BytesReference bytes = serializedDiffs.get(node.version());
+        if (bytes == null) {
+            try {
+                bytes = serializeDiffClusterState(diff, node.version());
+                serializedDiffs.put(node.version(), bytes);
+            } catch (Throwable e) {
+                logger.warn("failed to serialize diff of cluster_state before publishing it to node {}", e, node);
+                publishResponseHandler.onFailure(node, e);
+            }
+        }
+        publishClusterStateToNode(clusterState, bytes, node, timedOutWaitingForNodes, publishTimeout, publishResponseHandler);
+    }
+
+    private void publishClusterStateToNode(final ClusterState clusterState, BytesReference bytes,
+                                           final DiscoveryNode node, final AtomicBoolean timedOutWaitingForNodes,
+                                           final TimeValue publishTimeout,
+                                           final BlockingClusterStatePublishResponseHandler publishResponseHandler) {
+        try {
+            TransportRequestOptions options = TransportRequestOptions.options().withType(TransportRequestOptions.Type.STATE).withCompress(false);
+            // no need to put a timeout on the options here, because we want the response to eventually be received
+            // and not log an error if it arrives after the timeout
+            transportService.sendRequest(node, ACTION_NAME,
+                    new BytesTransportRequest(bytes, node.version()),
+                    options, // no need to compress, we already compressed the bytes
+
+                    new EmptyTransportResponseHandler(ThreadPool.Names.SAME) {
+
+                        @Override
+                        public void handleResponse(TransportResponse.Empty response) {
+                            if (timedOutWaitingForNodes.get()) {
+                                logger.debug("node {} responded for cluster state [{}] (took longer than [{}])", node, clusterState.version(), publishTimeout);
+                            }
+                            publishResponseHandler.onResponse(node);
+                        }
+
+                        @Override
+                        public void handleException(TransportException exp) {
+                            if (exp.unwrapCause() instanceof IncompatibleClusterStateVersionException) {
```

Comment:
can we add a check for whether we sent a diff? I want to avoid a potential infinite loop.